File: ObjC/LoadFromXIB/LoadFromXIB/LFXView.m
//
//  LFXView.m
//  LoadFromXIB
//
//  Created by Don Mag on 3/17/17.
//  Copyright © 2017 DonMag. All rights reserved.
//

#import "LFXView.h"

@interface LFXView ()

@property (weak, nonatomic) IBOutlet NSLayoutConstraint *constraintViewWidth;

@end

@implementation LFXView

- (id)initWithFrame:(CGRect)frame
{
	self = [super initWithFrame:frame];
	
	if (self) {
		[self xibSetup];
	}
	
	return self;
}

- (id)initWithCoder:(NSCoder *)coder
{
	self = [super initWithCoder:coder];
	
	if (self) {
		[self xibSetup];
	}
	
	return self;
}

- (void)awakeFromNib {
	[super awakeFromNib];
	[self xibSetup];
}

- (void)prepareForInterfaceBuilder {
	[self xibSetup];
}

- (void)xibSetup {
	
	if (!self.subviews.count) {
		
		UIView *view = [self loadFromXIB];
		
		view.frame = self.bounds;
		view.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
		view.layer.cornerRadius = 24.0;
		view.layer.masksToBounds = YES;
		[self addSubview:view];
		[self sendSubviewToBack:view];
		
		for (UIView *v in view.subviews) {
			if ([v isKindOfClass:[UILabel class]]) {
				self.theLabel = (UILabel *)v;
				break;
			}
		}
		
		// just an example of changing a constant when the custom view is loaded
		if (CGRectGetHeight([UIScreen mainScreen].bounds) == 667) {
			_constraintViewWidth.constant = 60;
		}
		
	}
	
}

- (UIView *)loadFromXIB {
	
	NSString *className = NSStringFromClass([self class]);
	NSBundle *bundle = [NSBundle bundleForClass:[self class]];
	UINib *nib = [UINib nibWithNibName:className bundle:bundle];
	UIView *v = [nib instantiateWithOwner:self options:nil].firstObject;
	
	return v;
	
}

- (void)setTheText:(NSString *)theText {
	self.theLabel.text = theText;
}


@end

File: Swift3/SW3LoadFromXIB/LFXView.swift
//
//  LFXView.swift
//  SW3LoadFromXIB
//
//  Created by Don Mag on 1/18/17.
//  Copyright © 2017 DonMag. All rights reserved.
//

import UIKit

@IBDesignable

class LFXView: UIView {

	@IBOutlet var theContentView: UIView!
	
	@IBOutlet weak var theLabel: UILabel!
	
	@IBInspectable var theText: String = "This is the XIB View" {
		didSet {
			self.theLabel.text = theText
		}
	}
	
	convenience init(with someText: String) {
		self.init(frame: CGRect.zero)
		theText = someText
		theLabel.text = someText
	}
	
	convenience init(with someText: String, and color: UIColor) {
		self.init(frame: CGRect.zero)
		theText = someText
		theLabel.text = someText
		theLabel.textColor = color
	}
	
	override init(frame: CGRect) {
		super.init(frame: frame)
		commonInit()
	}
	
	required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
		commonInit()
	}
	
	private func commonInit() {
		
		if self.subviews.count == 0 {
			
			let nib = UINib(nibName: String(describing: type(of: self)), bundle: Bundle(for: type(of: self)))
			if let cView = nib.instantiate(withOwner: self, options: nil).first as? UIView {
				theContentView = cView
			} else {
				// shouldn't happen
				theContentView = UIView()
			}
			
			addSubview(theContentView)
			
			theContentView.translatesAutoresizingMaskIntoConstraints = false
			
			theContentView.topAnchor.constraint(equalTo: topAnchor).isActive = true
			theContentView.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
			theContentView.leftAnchor.constraint(equalTo: leftAnchor).isActive = true
			theContentView.rightAnchor.constraint(equalTo: rightAnchor).isActive = true
			
			// just to show some customization
			theContentView.layer.cornerRadius = 24.0
			theContentView.layer.masksToBounds = true
			
		}
		
	}

}
